src/rail_gui.cpp

```diff
@@ -762,6 +762,45 @@ struct BuildRailToolbarWindow : Window {
 	static HotkeyList hotkeys;
 };
 
+/**
+ * Selects RailType based on _last_built_railtype and global modifiers from RailToolbarWidgetModifiers
+ *
+ * If GHK_MOD_* modifiers were used hotkey replaced with last selected widget
+ * or WID_RAT_AUTORAIL (if toolbar was closed).
+ *
+ * Without wrapping, so if you're on RAILTYPE_RAIL you'll stay there. Same for upper boundary.
+ *
+ * @param [out] hotkey Replaced by last user selected widget or WID_RAT_AUTORAIL if prev/next railtype used
+ * @return selected RailType
+ */
+static RailType ChooseRailTypeOnGlobalHotkey(int &hotkey)
+{
+	extern RailType _last_built_railtype;
+	RailType rt = _last_built_railtype;
+	bool change_rt = false;
+	if (HasBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE)) {
+		ClrBit(hotkey, GHK_MOD_RAT_PREV_RAILTYPE);
+		change_rt = true;
+		do {
+			rt--;
+		} while (rt > RAILTYPE_BEGIN && !HasRailtypeAvail(_local_company, rt));
+	} else if (HasBit(hotkey, GHK_MOD_RAT_NEXT_RAILTYPE)) {
+		ClrBit(hotkey, GHK_MOD_RAT_NEXT_RAILTYPE);
```

Comment:
A lot of duplicated code here, only real difference between these branches is loop direction. Pretty sure it can be implemented much simpler. At the very least you can get rid of change_rt (and ClrBit ofc).

Reply:
I'll think on it. Originally I did it to avoid complex and error-prone logic conditions. If I can reduce it to something simple and easy to follow -- ok)